### PR TITLE
Intrinsic plot sizing

### PR DESCRIPTION
**Pull request by seeM**
This PR adds a new sizing policy, the intrinsic sizing policy, which renders the plot at its "intrinsic" size i.e. the original size of the plot known to the backend.

Addresses https://github.com/posit-dev/positron/issues/4040.

Here's the corresponding Ark PR: https://github.com/posit-dev/ark/pull/470.

### How it works

1. Creating a plot with a backend that supports this feature, will add a new sizing policy. For example, in Python:

    ```python
    import matplotlib.pyplot as plt
    plt.subplots(figsize=(3, 10))[1].plot([0, 1])
    ```

2. The sizing policy's label will contain the source of the sizing information and the original size in the backend's preferred unit of measurement:

    image

3. You can also save plots at their intrinsic size:
    
    image

4. If the backend does not support intrinsic sizing, the policy's label will be disabled:

    image

    1. If you switch from a plot with the intrinsic sizing policy selected to a plot that does not support intrinsic sizing, it will fall back to the "auto" sizing policy.
    2. The "Use intrinsic sizing" checkbox will also be disabled in the save modal.

### QA Notes

1. Create Python plots with different sizes and make sure that switching between them updates the plot size and the sizing policy's label, e.g:
    ```python
    import matplotlib.pyplot as plt
    plt.subplots(figsize=(10, 3))[1].plot([0, 1])
    plt.subplots(figsize=(3, 10))[1].plot([0, 1])
    plt.subplots(figsize=(3, 4))[1].plot([0, 1])
    plt.subplots(figsize=(4, 3))[1].plot([0, 1])
    plt.subplots()[1].plot([0, 1])
    ```
2. Switching from a Python plot with the intrinsic sizing policy selected to an R plot should reset to the auto sizing policy.

### CODE REVIEW

**Comment by seeM** [comment]
Smoke test is failing since https://github.com/posit-dev/ark/pull/470 is required.

File: src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts
Comment (nstrayer) [inline src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts:548]: This is much nicer!

**Review by nstrayer — DISMISSED** [review]
This is awesome (and herculean). 

I don't have enough experience with the coms side of things to speak to best practices but everything looks good to me. I also added some notes about typescript/react patterns. Feel free to pick and choose what you want to take from it, nothing was a deal breaker.

File: positron/comms/plot-backend-openrpc.json
Comment (seeM) [reply positron/comms/plot-backend-openrpc.json:106]: This one is still in pixels, and `intrinsic_size` is in `plot_unit`.

File: positron/comms/data_explorer-backend-openrpc.json
Comment (seeM) [reply positron/comms/data_explorer-backend-openrpc.json:50]: Yuup. These two changes are because `generate-comms.ts` wasn't previously handling this field so all of our code assumes that these are actually required.

File: src/vs/workbench/contrib/positronPlots/browser/components/dynamicPlotInstance.tsx
Comment (seeM) [reply src/vs/workbench/contrib/positronPlots/browser/components/dynamicPlotInstance.tsx:46]: Good point! I'm happy to merge this after and fix conflicts here.

File: src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx
Comment (seeM) [reply src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx:263]: Ah good point. I've just removed the logs for now.

File: src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx
Comment (seeM) [reply src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx:396]: Good point, will do!

**Comment by seeM** [comment]
Thanks for the review @nstrayer! I've addressed all of your comments. Just need to merge the Ark PR, bump Ark, rerun the smoke tests, before we can merge.

File: src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx
Comment (jmcphers) [inline src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx:210]: Should this error be displayed to the user?

File: src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts
Comment (jmcphers) [inline src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts:284]: It's possible for this to be called quickly in succession since it's invoked during render. To avoid the possibility of stacking intrinsic size requests, we should probably check for an in-flight request here. Maybe use a `DeferredPromise` we can return?

File: positron/comms/data_explorer-backend-openrpc.json
Comment (jmcphers) [reply positron/comms/data_explorer-backend-openrpc.json:50]: I also wish JSON could have comments. :-(

File: src/vs/workbench/services/positronPlots/common/sizingPolicyIntrinsic.ts
Comment (jmcphers) [inline src/vs/workbench/services/positronPlots/common/sizingPolicyIntrinsic.ts:38]: This should be a localized string with placeholders (units should also be localized)

**Review by jmcphers — DISMISSED** [review]

File: src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx
Comment (seeM) [reply src/vs/workbench/contrib/positronPlots/browser/modalDialogs/savePlotModalDialog.tsx:210]: Yes!

File: src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts
Comment (seeM) [reply src/vs/workbench/services/languageRuntime/common/languageRuntimePlotClient.ts:284]: Good point, will do. Is there any reason to prefer a `DeferredPromise` over a `Promise` e.g. like:

```typescript
	public getIntrinsicSize(): Promise<IntrinsicSize | undefined> {
		// If there's already an in-flight request, return its response.
		if (this._currentIntrinsicSize) {
			return this._currentIntrinsicSize;
		}
		this._currentIntrinsicSize = this._comm.getIntrinsicSize()
			.then((intrinsicSize) => {
				this._intrinsicSize = intrinsicSize;
				this._receivedIntrinsicSize = true;
				this._didSetIntrinsicSizeEmitter.fire(intrinsicSize);
				return intrinsicSize;
			})
			.finally(() => {
				this._currentIntrinsicSize = undefined;
			});
		return this._currentIntrinsicSize;
	}
```

I've tended to avoid `DeferredPromise` where possible since I tend to forget to try-catch and reject properly.

**Pull request merged**
